### PR TITLE
Add target update capability to API

### PR DESCRIPTION
**Pull request by rcutmore**
This is an idea for fixing #18.

Currently we check all new targets against existing targets to see if any have been removed. If an existing target is missing from the list of new targets it gets removed from the database. This PR changes _PutGroup_ to update a target's information in the database if it still exists during this check.

Note: the web front end still needs to be updated to allow editing. This just adds the capability to the back end.

Let me know if I missed anything or if there's a different approach you'd prefer.

### CODE REVIEW

File: models/group.go [hunk 1]
```diff
@@ -222,6 +223,20 @@ func insertTargetIntoGroup(t Target, gid int64) error {
 	return nil
 }
 
+// UpdateTarget updates the given target information in the database.
+func UpdateTarget(target Target) error {
+	targetInfo := map[string]interface{}{
```
Comment (jordan-wright) [inline models/group.go]: Does this need to be a `map[string]interface`? can it just be a `models.Target` instance?

File: models/models_test.go [hunk 2]
```diff
@@ -63,6 +63,31 @@ func (s *ModelsSuite) TestPostGroupNoTargets(c *check.C) {
 	c.Assert(err, check.Equals, ErrNoTargetsSpecified)
 }
 
+func (s *ModelsSuite) TestPutGroup(c *check.C) {
```
Comment (jordan-wright) [inline models/models_test.go]: Thanks for adding tests! You're the hero we need, not the one we deserve 😄

**Comment by jordan-wright** [comment]
Looking good! Thanks for taking this on. I've added one note regarding the need for a `map[string]interface`. If it turns out we need that, that's totally cool. I just figured it might work with a `models.Target` instance.

File: models/group.go (hunk 1 above)
Comment (rcutmore) [inline models/group.go]: The docs show [here](http://jinzhu.me/gorm/curd.html#update) that a `map[string]interface` must be used to update all changed attributes, even empty ones. I tested it out using a `models.Target` instance and that seems to be the case.

I just added another test for this situation. If you update [this line](https://github.com/rcutmore/gophish/blob/92f4ca6a90197dda9c1289a68656c7b41ace57fc/models/group.go#L233) to `err := db.Model(&target).Where("id = ?", target.Id).Updates(target).Error` the new test will fail.

**Comment by rcutmore** [comment]
I added another test related to updating empty attributes (it seems `map[string]interface` is needed for this case).

Just a side note, it looks like there's no clean up of the test database between tests. I'm new to Go and not familiar with this test framework yet. For the new test I just created a new group with new targets to avoid conflicts with other tests, but I'm not sure if there's any convention regarding clean up between tests to avoid this.

**Comment by jordan-wright** [comment]
Thanks for knocking this out! I'll give it a shot over the next day or so and will merge if all looks good.

You're right - there currently isn't any database cleanup... I wasn't sure the best way to manage that when setting up the tests in the past. If you have any suggestions, let me know!

You're the man!

**Comment by jordan-wright** [comment]
Dude, awesome work again - I dig it. Merged!
